### ISSUE TITLE
Malloc error when merge interacts with dropping column family(java-jni)

### ISSUE THREAD

**Issue by scottfines** [facebook/rocksdb#2532]
I'm operating with Rocksdb v 5.4.5 through the RocksJava interface, and continue to run into a malloc error which causes my jvm to abort (SIGABRT). The simplest reproducing test case I have is as follows

```java
Charset charset = Charset.forName("UTF-8");
File tempFolder = File.createTempFile("mergeAndDrop",".");
tempFolder.delete();
tempFolder.mkdir();

try (final Options options = new Options() .setCreateIfMissing(true);
       final RocksDB db = RocksDB.open(options, tempFolder.getAbsolutePath())
	) {
	//add some data
	byte[] key = "hello".getBytes(charset);
	StringAppendOperator countOperator = new StringAppendOperator();
	for (int i = 0; i < 100; i++) {
		ColumnFamilyHandle cfHandle;
		try (final ColumnFamilyOptions cfOptions = new ColumnFamilyOptions().setMergeOperator(countOperator)){

			String cfName = "C_"+i;
			ColumnFamilyDescriptor cfDesc = new ColumnFamilyDescriptor(cfName.getBytes(charset), cfOptions);
			cfHandle = db.createColumnFamily(cfDesc);

		}

		byte[] value = ("happy happy happy").getBytes(charset);
		db.merge(cfHandle, key, value);

		//now drop the cf and repeat
		db.dropColumnFamily(cfHandle);
		System.out.printf("...");
		cfHandle.close();
	}
}
```

The error that is output is:

```
java(61986,0x70000993d000) malloc: *** error for object 0x7faf8d71ea90: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
```

This error does not occur at the same iteration each time--sometimes it will run for 10 iterations before failing, sometimes it fails the first time around. It has never gotten past 19 iterations before failing, however.

I've tried a number of variations on this theme, and here are some interesting (or not-so-interesting, perhaps) factoids:

1. The appending string in the example above ("happy happy happy") is the SHORTEST string under which I can reproduce the error; "happy happy happ" will work just fine, no matter how many iterations are chosen.

2. I tried several experiments to see if the error was related:
   A. Flushed CF before dropping
   B. Compact CF before dropping
   C. Slept for 1 second between the merge() call and the drop
   D. slept for 1 second after drop call (before next iteration)
   E. calling pauseBackgroundWork() before the loop
   F. used WriteOptions.setSync(true) to ensure merge was synchronous
   G. Disabled WAL for merge
   
None of which prevented the error from occurring; this is probably not surprising, but I checked them anyway, just to be sure.

The next step is probably to write a native version of the same loop and see if it breaks; alas, my C-fu is not very good, so I leave it to more skilled people than I :)

Platform:
Mac OS X(Sierra, version 10.12.5)
rocksdb version: 5.4.5, jnilib pulled from 5.4.5 jar in maven repo.
Java version: Hotspot 1.8.0_111

Let me know if you need more information/want me to re-run the test with different parameters

**Comment by adamretter** [comment]
The first issue that I notice is that - you should not close the `ColumnFamilyOptions` until after you have finished with and closed the `ColumnFamilyHandle`. Can you refactor for that and let me know if it solves your problem?

**Comment by adamretter** [comment]
@scottfines To make things clearer I have added to the Wiki on how the memory management for RocksJava should be taken care of. Please see:
1. https://github.com/facebook/rocksdb/wiki/RocksJava-Basics#memory-management
2. https://github.com/facebook/rocksdb/wiki/RocksJava-Basics#opening-a-database-with-column-families

**Comment by scottfines** [comment]
Thanks! the extra documentation is always helpful for schmoes like me :)

I tried this version of the test as well:

```
Charset charset = Charset.forName("UTF-8");
File tempFolder = File.createTempFile("mergeAndDrop",".");
tempFolder.delete();
tempFolder.mkdir();

try (final Options options = new Options() .setCreateIfMissing(true);
       final RocksDB db = RocksDB.open(options, tempFolder.getAbsolutePath())
	) {
//add some data
	byte[] key = "hello".getBytes(charset);
	byte[] value = ("happy happy happy").getBytes(charset);
	StringAppendOperator countOperator = new StringAppendOperator();
	for (int i = 0; i < 100; i++) {
		try (final ColumnFamilyOptions cfOptions = new ColumnFamilyOptions().setMergeOperator(countOperator)) {

			String cfName = "C_" + i;
			ColumnFamilyDescriptor cfDesc = new ColumnFamilyDescriptor(cfName.getBytes(charset), cfOptions);
			try(ColumnFamilyHandle cfHandle = db.createColumnFamily(cfDesc)) {

				db.merge(cfHandle, key, value);

				//now drop the cf and repeat
				db.dropColumnFamily(cfHandle);
			}
		}
	}
}
```

And still got the same error:

```
java(6528,0x70000d794000) malloc: *** error for object 0x7f9e79c27870: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
```

So it appears that bad management of CF options is not the culprit in this case.

**Comment by scottfines** [comment]
For kicks and giggles, just to see if manual resource closing was a problem, I also stripped it all out (obviously not a good idea in production):

```
for (int i = 0; i < 100; i++) {
	final ColumnFamilyOptions cfOptions = new ColumnFamilyOptions().setMergeOperator(countOperator);

	String cfName = "C_" + i;
	ColumnFamilyDescriptor cfDesc = new ColumnFamilyDescriptor(cfName.getBytes(charset), cfOptions);

	ColumnFamilyHandle cfHandle = db.createColumnFamily(cfDesc);

	db.merge(cfHandle, key, value);

	//now drop the cf and repeat
	db.dropColumnFamily(cfHandle);
}
```

and still got the malloc error. 

AFAICT, there aren't any resources being manually closed; perhaps Java GC was a problem? To check, I re-ran my test with `-XX:+PrintGC` enabled, but there were no logged GC events before the error occurred. This leads me to believe that whatever resources are being deleted are doing so outside of the JVM.

**Comment by scottfines** [comment]
I should also note that I periodically will get a SIGABRT error as well (with all forms of the test):

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000010ae86fa2, pid=6810, tid=0x0000000000001c03
#
# JRE version: Java(TM) SE Runtime Environment (8.0_111-b14) (build 1.8.0_111-b14)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.111-b14 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [librocksdbjni6526800766289161224.jnilib+0x2dffa2]  rocksdb::WritableFileWriter::Append(rocksdb::Slice const&)+0x172
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/scott/workspace/stardog/hs_err_pid6810.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

Process finished with exit code 134 (interrupted by signal 6: SIGABRT)
```

the associated error log is attached:

[hs_err_pid6810.txt](https://github.com/facebook/rocksdb/files/1133181/hs_err_pid6810.txt)

( I had to rename it to a .txt to get github to allow it)

**Comment by adamretter** [comment]
@scottfines I am afraid that I cannot seem to reproduce your issue. Using RocksDB 5.5.1 with your second code example with minimal modifications (I increased the number of iterations):

```java
import org.junit.Test;

import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Path;
import java.nio.file.Files;

public class ScottFinesTest {

    @Test
    public void test1() throws IOException, RocksDBException {

        final Charset charset = Charset.forName("UTF-8");
        final Path tempFolder = Files.createTempDirectory("mergeAndDrop");

        try (final Options options = new Options().setCreateIfMissing(true);
             final RocksDB db = RocksDB.open(options, tempFolder.toAbsolutePath().toString())
        ) {
            System.out.println("Opened database at: " + tempFolder.toAbsolutePath().toString());

            final byte[] key = "hello".getBytes(charset);
            final byte[] value = ("happy happy happy").getBytes(charset);

            final StringAppendOperator countOperator = new StringAppendOperator();
            for (int i = 0; i < 1000; i++) {
                try (final ColumnFamilyOptions cfOptions = new ColumnFamilyOptions().setMergeOperator(countOperator)) {

                    final String cfName = "C_" + i;
                    final ColumnFamilyDescriptor cfDesc = new ColumnFamilyDescriptor(cfName.getBytes(charset), cfOptions);
                    try (final ColumnFamilyHandle cfHandle = db.createColumnFamily(cfDesc)) {

                        System.out.println("mergeAndDrop on: " + cfName);

                        db.merge(cfHandle, key, value);

                        //now drop the cf and repeat
                        db.dropColumnFamily(cfHandle);
                    }
                }
            }
        }

    }

}
```

I have tested this on two platforms without issue:

1. macOS Sierra (10.12.5) / Oracle JDK 1.7.0_80-b15 and also Oracle JDK 1.8.0_131.

2. Ubuntu Linux 17.04 / IBM JDK 1.7.0 (J9 VM build 2.7)

Can you test with RocksDB 5.5.1 please?

**Comment by scottfines** [comment]
Sorry, just saw the last comment. Will try with latest release in public repo (5.6.1) and 5.5.1.

**Comment by scottfines** [comment]
Looks like this is resolved in 5.6.1! Maybe a regression on 5.4.5? It definitely still occurs on the 5.4.5 release.

So the recommended resolution here is probably to just avoid the 5.4.5 release if you are using merge through JNI?

**Comment by adamretter** [comment]
@scottfines I would certainly recommend using the latest RocksJava.
